core/src/main/java/io/grpc/util/CertificateUtils.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2021 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.util;
+
+import com.google.common.base.Charsets;
+import com.google.common.io.BaseEncoding;
+import com.google.common.io.CharStreams;
+import io.grpc.ExperimentalApi;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.security.KeyException;
+import java.security.KeyFactory;
+import java.security.PrivateKey;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.util.Collection;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Contains certificate/key PEM file utility method(s).
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/8024")
+public final class CertificateUtils {
+  private static CertificateFactory factory;
+  private static final Pattern KEY_PATTERN = Pattern.compile(
+      "-+BEGIN\\s+.*PRIVATE\\s+KEY[^-]*-+(?:\\s|\\r|\\n)+" + // Header
```

Comment:
This looks _far_ too generous. See https://tools.ietf.org/html/rfc7468#section-2

Seems the line must be a very specific string, without any variation. (Number of dashes and spaces are all strictly defined.) At that point there doesn't seem to be much point in using a regex here. Can we use BufferedReader instead?

Reply:
Yeah, this regex pattern(which was copied from Netty's PEM key reader implementation) seems not strict enough...I made the regex a bit more strict.
Regarding to using `BufferedReader`, do you want me to read line by line, and check the header and footer respectively? That seems fine, but since we don't know when we are at the end line, we will need two variables to control it. That makes the code less cleaner from my point of view.
Do we really want that change? Or do you have any other suggestions here? Thanks in advance!

Reply:
> Regarding to using BufferedReader, do you want me to read line by line, and check the header and footer respectively?

Reading line-by-line is what I expected.

> That seems fine, but since we don't know when we are at the end line, we will need two variables to control it

I don't understand. I expected you to use `readLine()`.

```java
String line;
while ((line = reader.readLine()) != null) {
 if ("-----BEGIN PRIVATE KEY-----".equals(line)) {
    break;
  }
}
StringBuilder base64Stuff = new StringBuilder();
while ((line = reader.readLine()) != null) {
 if ("-----END PRIVATE KEY-----".equals(line)) {
    break;
  }
  base64Stuff.add(line);
}
...
```

Reply:
Yeah, I think this looks better than I thought. Updated to using `BufferedReader`.